controllers/quay/quayregistry_controller.go

```diff
@@ -120,10 +120,108 @@ func (r *QuayRegistryReconciler) Reconcile(ctx context.Context, req ctrl.Request
+		if err != nil && errors.IsNotFound(err) {
+			_, err = r.updateWithCondition(
+				updatedQuay,
+				v1.ConditionComponentsCreated,
+				metav1.ConditionFalse,
+				v1.ConditionReasonMigrationsJobMissing,
+				"upgrade job not found")
+			if err != nil {
+				log.Error(err, "failed to update `conditions` of `QuayRegistry`")
+			}
+			return ctrl.Result{RequeueAfter: time.Second}, nil
+		}
+
+		if err != nil {
+			log.Error(err, "could not retrieve Quay upgrade `Job`")
+			return ctrl.Result{RequeueAfter: time.Second}, nil
+		}
+
+		if upgradeJob.Status.Active == 1 {
+			return ctrl.Result{RequeueAfter: time.Second}, nil
+		}
+
+		if upgradeJob.Status.Succeeded == 1 {
+			log.Info("Quay upgrade complete, updating `status.currentVersion`")
+
+			var freshQuay v1.QuayRegistry
+			if err := r.Client.Get(ctx, req.NamespacedName, &freshQuay); err != nil {
+				log.Error(err, "could not retrieve QuayRegistry")
+				return ctrl.Result{RequeueAfter: time.Second}, nil
+			}
+			qcopy := freshQuay.DeepCopy()
+
+			msg := "All registry components created"
+			condition := v1.Condition{
+				Type:               v1.ConditionComponentsCreated,
+				Status:             metav1.ConditionTrue,
+				Reason:             v1.ConditionReasonComponentsCreationSuccess,
+				Message:            msg,
+				LastUpdateTime:     metav1.Now(),
+				LastTransitionTime: metav1.Now(),
+			}
+			qcopy.Status.Conditions = v1.SetCondition(qcopy.Status.Conditions, condition)
+			qcopy.Status.CurrentVersion = v1.QuayVersionCurrent
+
+			if err = r.Client.Status().Update(ctx, qcopy); err != nil {
+				log.Error(err, "could not update QuayRegistry status with current version")
+				return ctrl.Result{RequeueAfter: time.Second}, nil
+			}
+
+			log.Info("successfully updated `status` after Quay upgrade")
+
+			return ctrl.Result{RequeueAfter: time.Second}, nil
+		}
+
+		// when the job fails, the next reconciliation to run will think
+		// that migrations are currently not running, since we change
+		// the condition reason to v1.ConditionReasonMigrationsFailed.
+		// this doesn't necessarily describe reality, because kube will
+		// retry failed jobs for us, but it's desired behaviour because
+		// when the migration job fails due to misconfiguration, then the
+		// reconcile function should be allowed to proceed.
+		if upgradeJob.Status.Failed == 1 {
+			msg := "failed to run migrations"
+			for _, cond := range upgradeJob.Status.Conditions {
+				if cond.Type == batchv1.JobFailed && cond.Status == corev1.ConditionTrue {
+					msg = cond.Message
+					break
+				}
+			}
+			updatedQuay, err = r.updateWithCondition(
+				updatedQuay,
+				v1.ConditionComponentsCreated,
+				metav1.ConditionFalse,
+				v1.ConditionReasonMigrationsFailed,
+				msg)
+			if err != nil {
+				log.Error(err, "failed to update `conditions` of `QuayRegistry`")
+			}
+			return ctrl.Result{RequeueAfter: time.Second}, nil
+		}
+
+		return ctrl.Result{RequeueAfter: time.Second}, nil
```

Comment:
We gonna hit here if `Active != 1 && Failed != 1 and Succeeded != 1`. I think that this should never happen so it could be useful to log something so we can see that we are under this condition. I am saying that because I really don't know if this condition could happen but better be safe than sorry. wdyt?

Reply:
agreed, we should never end up here. I'll come up with something to log.